llvm/lib/CodeGen/RegisterCoalescer.cpp

```diff
@@ -3672,6 +3672,14 @@ bool RegisterCoalescer::joinVirtRegs(CoalescerPair &CP) {
     // having stale segments.
     LHSVals.pruneMainSegments(LHS, ShrinkMainRange);
 
+    LHSVals.pruneSubRegValues(LHS, ShrinkMask);
+    RHSVals.pruneSubRegValues(LHS, ShrinkMask);
+  } else if (TrackSubRegLiveness && !CP.getDstIdx() && CP.getSrcIdx()) {
```

Comment:
Add comment like "MachineVerifier expects subrange to be created if a register is partially defined when subregliveness is enabled"?

Reply:
If CP has a src idx, I would expect the source value to have subranges when `TrackSubRegLiveness` is on.
Bottom line, I don't see why we need this case and thus echoing what @arsenm said wrt adding a test.

Reply:
IIRC, it has something to do with `SUBREG_TO_REG` in which neither Src nor Dst has subranges.

Reply:
SUBREG_TO_REG coalescing is pretty broken. It would be great if you could migrate to not using it at all. It's fundamentally flawed and I think should be removed

Reply:
I'm not sure we can achieve it in short term. Grep a bit, PowerPC has the most lines of code involving `SUBREG_TO_REG` comparing to X86 and AArch64.

Reply:
> I'm not sure we can achieve it in short term. Grep a bit, PowerPC has the most lines of code involving `SUBREG_TO_REG` comparing to X86 and AArch64.

I agree we should be moving away from `SUBREG_TO_REG` however I also agree that for PowerPC this is more of a long term goal. For this patch we will probably have to continue like this.

Reply:
Is it possible to split this change and test into its own PR

Reply:
> Is it possible to split this change and test into its own PR

The issue with splitting this test and change away is that the test passes anyway unless it is accompanied by the subregsiter changes in this PR.  So, unfortunately, if I split out the test and add it to top of trunk main the test will PASS even without the changes to `RegisterCoalescer.cpp`. The error is exposed by the subregister changes that we made in `PPCRegisterInfo.td`.

Reply:
That's fine, pre-committing passing tests is good